dask/highlevelgraph.py

```diff
@@ -706,7 +706,23 @@ def cull(self, keys: Iterable) -> HighLevelGraph:
 
         keys_set = set(flatten(keys))
 
-        all_ext_keys = self.get_all_external_keys()
+        class InGraph:
```

Comment:
I like the idea here, but propose the following:

- We pull this out of the `HighLevelGraphClass` (we typically try to avoid nested classes like this in dask source code)
- We call this something like `LazyKeySet` (Although, I don't feel strongly about this. `InGraph` just feels "off" for some reason)

Therefore, the class definition would look something like the following:

```python
class LazyKeySet:
    """Wraps a ``HighLevelGraph`` object to provide a lazy view of its keys
    
    This class must be compatible with ``Layer.cull``.
    """
    def __init__(self, graph: HighLevelGraph):
        self._graph = graph

    def __contains__(self, key):
        return any(key in layer for layer in self._graph.layers.values())
    ...
```

Reply:
Sure, that seems perfectly reasonable to me. Fundamentally a separate class isn't even necessary at all, as we could just pass the graph directly to Layer.cull and define the `__contains__` method on the HighLevelGraph class itself. 

This seems to more or less align with what HighLevelGraph.__getitem__ is doing anyway as far as key membership is concerned in the final O(n) access block. We would just need to include a check of `(key in self.layers and key in self.layers[key]) or (key[0] in self.layers and key in self.layers[key[0]])` to fully replicate the check.